R/extract.R

```diff
@@ -98,8 +98,25 @@ filter_chunk_end_positions <- function(starts, ends) {
   code_ends
 }
 
+bookdown_engines <- c(
```

Comment:
does bookdown export anything we could use instead of maintaining this?

Reply:
Nothing I could find, unfortunately. Should we request it?

But I'd prefer if this doesn't block the current PR.

Reply:
Yes, please file an issue, and mark a comment here referring to the "upstream" issue #

Reply:
Actually, looks like the additional engines are registered automatically if the library is loaded, but we don't want to gain dependence on `{bookdown}`, so not sure what to do here. 

Should I still create an issue?

``` r
sort(names(knitr::knit_engines$get()))
#>  [1] "asis"      "asy"       "awk"       "bash"      "block"     "block2"   
#>  [7] "bslib"     "c"         "cat"       "cc"        "coffee"    "comment"  
#> [13] "css"       "ditaa"     "dot"       "embed"     "eviews"    "exec"     
#> [19] "fortran"   "fortran95" "gawk"      "glue"      "glue_sql"  "gluesql"  
#> [25] "go"        "groovy"    "haskell"   "highlight" "js"        "julia"    
#> [31] "lein"      "mysql"     "node"      "octave"    "perl"      "php"      
#> [37] "psql"      "python"    "R"         "Rcpp"      "Rscript"   "ruby"     
#> [43] "sas"       "sass"      "scala"     "scss"      "sed"       "sh"       
#> [49] "sql"       "stan"      "stata"     "targets"   "tikz"      "verbatim" 
#> [55] "zsh"

library(bookdown)
sort(names(knitr::knit_engines$get()))
#>  [1] "asis"        "asy"         "awk"         "bash"        "block"      
#>  [6] "block2"      "bslib"       "c"           "cat"         "cc"         
#> [11] "coffee"      "comment"     "conjecture"  "corollary"   "css"        
#> [16] "definition"  "ditaa"       "dot"         "embed"       "eviews"     
#> [21] "example"     "exec"        "exercise"    "fortran"     "fortran95"  
#> [26] "gawk"        "glue"        "glue_sql"    "gluesql"     "go"         
#> [31] "groovy"      "haskell"     "highlight"   "hypothesis"  "js"         
#> [36] "julia"       "lein"        "lemma"       "mysql"       "node"       
#> [41] "octave"      "perl"        "php"         "proof"       "proposition"
#> [46] "psql"        "python"      "R"           "Rcpp"        "remark"     
#> [51] "Rscript"     "ruby"        "sas"         "sass"        "scala"      
#> [56] "scss"        "sed"         "sh"          "solution"    "sql"        
#> [61] "stan"        "stata"       "targets"     "theorem"     "tikz"       
#> [66] "verbatim"    "zsh"
```

<sup>Created on 2022-09-24 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
Does that mean the issue is solved by running `library(bookdown)` before running the linter?

Reply:
Probably, but we shouldn't count on it, since it's unusual to explicitly load this library in `{bookdown}` projects. 

For static analysis, it would just be easier to have a vector of supported engines, like we are doing for `{knitr}`. Note that users don't need to do `library(knitr)`, so it'd be surprising for them to have to do so for `{bookdown}`.

Reply:
Hmm, still not a fan of maintaining a long list of things which looks likely to change in the future (and the list likely to grow for other knitr engines), at least we should be getting the names supplied by bookdown:

https://github.com/rstudio/bookdown/blob/7e9b46fee109f06f86962d6a78b801d3491c94aa/R/utils.R#L551-L556

https://github.com/rstudio/bookdown/blob/dfe92a227b1dce5b8f0a2f758fb545e7df1b7e62/R/zzz.R#L5

> Note that users don't need to do library(knitr), so it'd be surprising for them to have to do so for {bookdown}.

That's not _too_ surprising IMO, `knitr` is the default / core engine for the Rmd format, and so _it_ is the exception rather than bookdown. All other engines would have the same advice: run `library(some.engine)` first to register the chunk engines; this rule also applies to `bookdown`.

Reply:
Okay, I have reverted this change to rely on `{bookdown}` being loaded.